From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Clayton Gilmer <claytongilmer24@outlook.com>
Date: Sat, 22 Apr 2023 15:20:59 -0400
Subject: [PATCH] 5252: SPARC: Add recognition for PC relative addressing

feat(SPARC): add injection for __sparc_get_pc_thunk

fix(SPARC): remove extraneous +'s from SPARC_patterns.xml
---
 Ghidra/Processors/Sparc/data/languages/SparcV9_32.cspec  | 9 +++++++++
 Ghidra/Processors/Sparc/data/languages/SparcV9_64.cspec  | 9 +++++++++
 Ghidra/Processors/Sparc/data/patterns/SPARC_patterns.xml | 5 +++++
 3 files changed, 23 insertions(+)

diff --git a/Ghidra/Processors/Sparc/data/languages/SparcV9_32.cspec b/Ghidra/Processors/Sparc/data/languages/SparcV9_32.cspec
index 28f5f3006a..76d613934f 100644
--- a/Ghidra/Processors/Sparc/data/languages/SparcV9_32.cspec
+++ b/Ghidra/Processors/Sparc/data/languages/SparcV9_32.cspec
@@ -100,4 +100,13 @@
       </localrange>
     </prototype>
   </default_proto>
+  <callfixup name="get_pc_thunk">
+    <target name="__sparc_get_pc_thunk"/>
+    <pcode>
+      <body><![CDATA[
+      l7 = inst_start + l7;
+      ]]>
+      </body>
+    </pcode>
+  </callfixup>
 </compiler_spec>
diff --git a/Ghidra/Processors/Sparc/data/languages/SparcV9_64.cspec b/Ghidra/Processors/Sparc/data/languages/SparcV9_64.cspec
index 67ccac7080..a388e578bd 100644
--- a/Ghidra/Processors/Sparc/data/languages/SparcV9_64.cspec
+++ b/Ghidra/Processors/Sparc/data/languages/SparcV9_64.cspec
@@ -187,4 +187,13 @@
         <range space="stack" first="0x0" last="0x8ae"/>   <!-- Stack bias of 7FF + 0xb0 window size -->
       </localrange>
     </prototype>
+  <callfixup name="get_pc_thunk">
+    <target name="__sparc_get_pc_thunk"/>
+    <pcode>
+      <body><![CDATA[
+      l7 = inst_start + l7;
+      ]]>
+      </body>
+    </pcode>
+  </callfixup>
 </compiler_spec>
diff --git a/Ghidra/Processors/Sparc/data/patterns/SPARC_patterns.xml b/Ghidra/Processors/Sparc/data/patterns/SPARC_patterns.xml
index 3d1976d23e..b097c25347 100644
--- a/Ghidra/Processors/Sparc/data/patterns/SPARC_patterns.xml
+++ b/Ghidra/Processors/Sparc/data/patterns/SPARC_patterns.xml
@@ -32,4 +32,9 @@
       <funcstart/>
     </postpatterns>
   </patternpairs>
+  <pattern>
+    <!-- retl; add o7, l7, l7 -->
+    <data>0x81 0xc3 0xe0 0x08 0xae 0x03 0xc0 0x17</data>
+    <funcstart label="__sparc_get_pc_thunk" validcode="function"/>
+  </pattern>
  </patternlist>
-- 
2.45.0

